presto-main/src/main/java/com/facebook/presto/operator/scalar/SplitToMultimapFunction.java

```diff
@@ -0,0 +1,117 @@
+    {
+        pageBuilder = new PageBuilder(ImmutableList.of(mapType));
+    }
+
+    @SqlType("map<varchar,array<varchar>>")
+    public Block splitToMultimap(
+            @TypeParameter("map<varchar,array<varchar>>") Type mapType,
+            @SqlType(StandardTypes.VARCHAR) Slice string,
+            @SqlType(StandardTypes.VARCHAR) Slice entryDelimiter,
+            @SqlType(StandardTypes.VARCHAR) Slice keyValueDelimiter)
+    {
+        checkCondition(entryDelimiter.length() > 0, INVALID_FUNCTION_ARGUMENT, "entryDelimiter is empty");
+        checkCondition(keyValueDelimiter.length() > 0, INVALID_FUNCTION_ARGUMENT, "keyValueDelimiter is empty");
+        checkCondition(!entryDelimiter.equals(keyValueDelimiter), INVALID_FUNCTION_ARGUMENT, "entryDelimiter and keyValueDelimiter must not be the same");
+
+        Multimap<Slice, Slice> multimap = ArrayListMultimap.create();
+        int entryStart = 0;
+        while (entryStart < string.length()) {
+            // Extract key-value pair based on current index
+            // then add the pair if it can be split by keyValueDelimiter
+            Slice keyValuePair;
+            int entryEnd = string.indexOf(entryDelimiter, entryStart);
+            if (entryEnd >= 0) {
+                keyValuePair = string.slice(entryStart, entryEnd - entryStart);
+            }
+            else {
+                // The rest of the string is the last possible pair.
+                keyValuePair = string.slice(entryStart, string.length() - entryStart);
+            }
+
+            int keyEnd = keyValuePair.indexOf(keyValueDelimiter);
+            if (keyEnd < 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Key-value delimiter must appear exactly once in each entry. Bad input: " + keyValuePair.toStringUtf8());
+            }
+
+            int valueStart = keyEnd + keyValueDelimiter.length();
+            Slice key = keyValuePair.slice(0, keyEnd);
+            Slice value = keyValuePair.slice(valueStart, keyValuePair.length() - valueStart);
+            if (value.indexOf(keyValueDelimiter) >= 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Key-value delimiter must appear exactly once in each entry. Bad input: " + keyValuePair.toStringUtf8());
+            }
+
+            multimap.put(key, value);
+
+            if (entryEnd < 0) {
+                // No more pairs to add
+                break;
+            }
+            // Next possible pair is placed next to the current entryDelimiter
+            entryStart = entryEnd + entryDelimiter.length();
+        }
+
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+        BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(0);
+        BlockBuilder singleMapBlockBuilder = blockBuilder.beginBlockEntry();
+        for (Map.Entry<Slice, Collection<Slice>> entry : multimap.asMap().entrySet()) {
```

Comment:
You can simplify this as
```java
multimap.forEach((key, values) -> {
    ...
});
```
`Map` has a similarly useful `forEach` method. The lambda is nice because you avoid the `Map.Entry<...>`  boilerplate and needing to unpack the key/value.